Unicode utils

fbacher opened the pull request:
## 19883: Switch much of StringUtils to use Unicode ICU library 
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
This is a large change. It is possible to break changes up and phase them in.
- Move Unicode dependent StringUtils functions to UnicodeUtils, to avoid confusion
- Some, but not many, changes to api or behavior
- More detail: 
[README UNICODE.md](https://github.com/xbmc/xbmc/files/9129811/README.UNICODE.md)

ICU library is already a dependency of several Kodi dependencies (see README.Unicode). Also mentioned in the README is that there is an open source project which provides CMake script/tools to download, build, integrate ICU with other projects. It is probably useful. Attached are the several bash scripts that I use. Simply modify bash_config.sh to specify your paths. The ICU ones are the only ones that are likely of any use to you. 
[scripts.zip](https://github.com/xbmc/xbmc/files/9130046/scripts.zip)

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
There are multiple Unicode issues with Kodi. The Turkish Locale/Turkic I problem are the worst. But  there are
numerous other problems caused by improper Unicode processing, very inadequate support in C++, treating UTF-8 as single byte characters, not realizing that it can take multiple Unicode codepoints (32 bits) to represent a character/grapheme, etc.
<!--- If it fixes an open issue, please link to the issue here. -->
https://github.com/xbmc/xbmc/issues?q=is%3Aopen+mentions%3A%40me+label%3A%22Triage%3A+Needed%22

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
Added more tests to TestStringUtils and TestUnicodeUtils. Ran all automated tests with 100% pass
Ran Kodi with English, Turkish, Ukranian, Russian, Hungarian, German, Japanese, Chinese, Iranian. I only know some German, but I was able to discover and resolve a number of issues.
It is impossible to test the entire UI, especially in languages you can't read.
Ran under lldb and with memory corruption tools enabled.
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Develop and test on Ubuntu Linux 21.10 ONLY. 
The core code is in only a few files, but TONS of code uses it. 
Most changes to other code is to just change the class name. 
Almost all uses of StringUtils.ToLower were changed to UnicodeUtils::FoldCase. This is the main fix for the Turkic I problem. (Note that not everyone's FoldCase will fix this issue. In particular, C++ version does NOT, Neither does Python's.)
This and the other changes generally don't require a lot of work. The hardest part is trying to determine if text is for internal processing or user presentation. (You don't want to FoldCase something to be presented to the user). Most of the time this is easy. If there is much guesswork, leave it alone and worry about it later.

I did make some changes that were not strictly necessary to fix the problem at hand:
Provided ICU Collator to use in place of Kodi's Legacy, English-centric StringUtils::AlphanumericComparision. The ICU version is highly configurable and it should closely mimic AlphanumericComparison. There is a #define to select which one to use. The new one should work much better in other locales.

Rather than blather on it is probably best to read my attached README.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
- Able to use Kodi in all languages
- General improvement in other languages due to eliminating numerous Unicode issues and using an industrial-strength ICU library instead of home-grown and C++.
- Besides the risk of the shear size of the change, the biggest negative risk is performance. I have not yet run performance tests. Unicode processing is slower and the approach that I took is by no means the most efficient. I focused on getting it working and experimenting with different approaches. Performance testing and work should be done prior to release.
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [x] I have added tests to cover my change
- [x] All new and existing tests passed

fbacher committed:
Added UnicodeSupport- Core changes

First of several interdependant commits

Moved Unicode dependent apis from StringUtils to UnicodeUtils so that it
is clear which ones use the ICU library and are Unicode safe. It also
gives a little latitude to api changes:
- Adding Locale argument
- Redefine "count" argument for Left/Mid/Right to mean characters and
not bytes. This doesn't impact current usage since they were all
assuming bytes, even if wrong.
- Add argument to Left/Mid/Right to indicate whether selected substring
defines what is to be returned, or omitted. Ex. Left("abc", 1, false)
says to omit the rightmost character, returning "ab"

fbacher committed:
Add ability to sort using ICU

Added call UnicodeUtils::InitializeCollator and SortCompleted so that
ICU collator can work.

The sort method is determined by UnicodeUtils::#define USE_ICU_COLLATOR.
The  current setting is ICU.
The Legacy Kodi collator is English centric. The ICU collator produces
very similar results and is locale sensitive and configurable.

fbacher committed:
Python support for query of locale settings

Added getICULanguage so that Python addons can get information for the
current local that getLanguage api can not give. The information comes
directly from the ICU library.

Also, fixed a bug in getLanguage. Other problems exist, but the code
scared me.

The bug in question is caused by calling ConvertToISO6391 twice when
region is specified. The second time with a raw region name, when
ConvertToISO6391 expects a lang.

fbacher committed:
Use new apis

fbacher committed:
Set default filename encoding to UTF-8

A side-effect of the 19.1 Turkish patch was to force default Python
filename encoding to ASCII. This required that every open on non-ASCII
filenames to specify UTF-8. This was bad enough for addons, but murder
for dependent code.

The root cause was changing Kodi's Locale to C, which is ASCII. The only
way to fix is by this change, or something similar. You can't make the
change after the initial startup of the Python Interpreter.

fbacher commented:
To simplify review of "Use new apis" you might modify a copy of master:
Change all instances of StringUtils to UnicodeUtils in the  tree, then diff against this PR.
This will change some functions that are still StringUtils to UnicodeUtils so you could rename those back if you like, such as Tokenize, Format. SecondsToTime

fbacher commented:
We could do this incrementally:
Choice 1: commit proposed UnicodeUtils.cpp, Unicode.cpp and ICU library, but incrementally modify StringUtils and rest of code to utilize UnicodeUtils
Choice 2: commit ICU library then incrementally switch code to use UnicodeUtils functions, adding the needed API to UnicodeUtils and removing from StringUtils

API priority:
1 Set Python's default filename encoding to UTF-8. This is small and would no longer require addons to explicitly set encodings for filenames, even if patch for 19883 remains. 
2 FoldCase- as far as I know, this is the biggest problem to fixing the problem of issue 21690. Most uses of ToLower would switch to FoldCase.
3 Perhaps as important as priority 2, switch all *NoCase functions to the UnicodeUtils version 
4 Add Python API getICULanguage so that addons can get vital info about current Locale. The existing getLanguage is not up to the task. Today an addon cannot get its language and region
2 CompareNoCase(s1, s2, size_t, ...) to use StartsWithNoCase & EndsWithNoCase. This eliminates some problems with splitting multi-byte characters in the middle by the size_t (length))
3 Left, Right, Mid The situation here is similar to CompareNoCase. Redefine count argument to mean characters, not bytes. Also add keepLeft and keepRight arguments per the new Left & Right apis.

I would have to review the APIs to see what the other priorities would be. 

Anyway, with this approach, it would spread the heavy lifting around. I do not know what platforms already have ICU as a dependency or readily available. So adding ICU might be a heavy lift. Also, the FoldCase change is fairly straight forward, but a lot of changes, making it fairly heavy. Finally, this would allow the UnicodeUtils code to be reworked for efficiency as we go along. 

I would prefer this big drop since I'm afraid of not getting all of the changes in this release. However, this is definitely a big change. Doing it incrementally makes it easier to isolate problems as they arise as well as focusing development (presumably me) on fixing those problems.

Question: if an incremental approach is taken, how many platforms would have to be converted over to use ICU before deploying further api changes? Is Python and Windows enough? Would we #ifdef USE_UNICODE_API around every call to a changed api? We may end up needing #ifdef USE_UNICODE_XXX_API with XXX being a function or phase name so that the different platforms could be migrated at different rates.

I only able to work with Linux and Windows (I first have to upgrade my Windows). I am most efficient with Linux.

fbacher commented:
How do I resolve check failures: Mergeable and default? I remember a PR naming convention for addons when I supplied a fix for an orphan addon. But I don't see any conventions specified in CONTRIBUTING guide.

fuzzard commented:
Mergable failures arent your issue to be concerned about.
Not a single platform actually builds. there is no current dependency on icu4c for any platform. That will need to be resolved for everything before you will gain any traction with the PR unfortunately.

garbear commented:
Mergeable fixed.

@fbacher It looks like you're bringing in a new dependency. As someone who's added a dependency (flatbuffers) I can relay that it's a long process, because we cover so many platforms, but if the dependency is worth it then there's lots of people to help with the edge cases once you've covered the major platforms.

fbacher commented:
On Linux Kodi dependencies  libsamba and libxml2 use ICU. They get it via libicu. At least for that platform we could do the same.

fbacher commented:
Products that use ICU library: https://icu.unicode.org/#h.f9qwubthqabj
It is ported to numerous platforms, but finding official binaries might be a challenge.  I'll dig into it more.

Windows supports it, but ONLY C apis, not C++. This may mean that we should build our own copy. : https://docs.microsoft.com/en-us/windows/win32/intl/international-components-for-unicode--icu-

> In Windows 10 Creators Update, ICU was integrated into Windows, making the C APIs and data publicly accessible.
> The version of ICU in Windows only exposes the C APIs. It does not expose any of the C++ APIs. Unfortunately, it is impossible to ever expose the C++ APIs due to the lack of a stable ABI in C++.

I don't see it officially available on macOS from Apple

wsnipex commented:
we have to build our own in any case

CastagnaIT commented:
what i mean with the use of std::string_view / std::wstring_view is that will optimize the performace because each time when you use std::string you are creating a new object and new allocation of the original data where string_view is a lightweight container
can be used when no specified things are required

I have not checked all the code changes, but example this:

```c++
bool UnicodeUtils::EqualsNoCase(const std::string &str1, const char *s2, StringOptions opt /* StringOptions::FOLD_CASE_DEFAULT */,
		const bool Normalize /* = false */) {
	std::string str2 = std::string(s2);
	return EqualsNoCase(str1, str2, opt, Normalize);
}

bool UnicodeUtils::EqualsNoCase(const char *s1, const char *s2, StringOptions opt /* = StringOptions::FOLD_CASE_DEFAULT */,
		const bool Normalize /* = false */) {
	std::string str1 = std::string(s1);
	std::string str2 = std::string(s2);
	return EqualsNoCase(str1, str2, opt, Normalize);
}
```

IMO as done is not so appropriate each time you are creating double allocations of same data, or also more times if similar things happens on sub-methods called

a better solution is something similar to this:

```c++
bool UnicodeUtils::EqualsNoCase(const std::string_view str1, const std::string_view str2, StringOptions opt /* StringOptions::FOLD_CASE_DEFAULT */,
		const bool Normalize /* = false */) {
	return EqualsNoCase(str1, str2, opt, Normalize);
}
```

one method that accept both inputs (char* and std::string), no need to kept both methods,
no new string objects and new allocations, the data is always referred to the original pointer,
this is purely a illustrative example

fbacher committed:
Jenkins formatting changes

Only kept changes to core files

fbacher committed:
Save work

Using string_view
Improved naming
Performance changes

CastagnaIT commented on xbmc/utils/UnicodeUtils.cpp:
```diff
@@ -53,598 +53,699 @@
 }
-int UnicodeUtils::CompareNoCase(const std::string &str1, const std::string &str2,
-    StringOptions opt /* = StringOptions::FOLD_CASE_DEFAULT */, const bool Normalize /* = false */) {
+int UnicodeUtils::CompareNoCase(const std::string_view& str1,
```
a better step but using as reference with `view_string` types (`std::string_view&`) no makes sense, the container use a const pointer and little else,
but worse you prevent the use of view_string with char arrays / std::string types
which is one of the goal to remove superfluous overload methods

Just a side note, add the left `const` to `string_view` has no particular application, locks only the methods of the viewstring container, this because the view itself is always const/readonly and will never influence the origin of the data, then add `const` is fine only when you don't have to use the view methods. You will probably never use these methods since you make use of all customised string operation methods so its fine

lrusak commented:
This is a challenge because of the really large diff involved (+13,958 −3,887) and that doesn't even include the required dependency changes.

I see a few things you could do to help the reviewing process (else this may never get reviewed):
 - Do everything through `StringUtils` to avoid the class change from `StringUtils` -> `UnicodeUtils`. You can always update the change set in another commit (or probably another PR to keep the initial changes low). That makes reviewing the changes easier and you can script the class name change (in case it needs to be done again).
 - Avoid changing method names `ToLower` -> `FoldCase` (at least in this PR). You can do this in a separate commit using a script (in case it needs to be done again).
 - The last tips boil down to making logical commits that are easily reviewable individually with a scoped set of changes.

You may also want to do the libicu dependency changes in a PR before doing any other changes to this PR.

You will also need cmake changes to find libicu and it's headers therwise the library won't be linked (I'm not sure how you built it locally or if it just works by luck).

Don't be afraid to ask for help 👍

fbacher commented:
I am incorporating the suggestions. 
I am also making changes to hopefully make easier to review. Little things like making boilerplate code follow the same pattern and naming conventions. Boilerplate code to do things like create stack buffers, convert to/from utf8, etc.

I'll open a separate PR for ICU. I have some bash scripts and made a primitive change to the top level CMakeLists. I'm no CMake wizard. There is an open source CMake wrapper for building ICU that I have used, but I don't know how to bolt it into our system.
This wrapper downloads, configures, builds and integrates the product into your build.

Questions
Several apis have signatures that take a single char instead of a string. Should I change the chars to strings and eliminate the signatures? This effects two apis, so not a big hit.

Also, I have not looked too carefully at how wstrings are used, but I understand that on Windows wchar_t is 16 bits, yet I see code the implies that char32 is copied to/from it. 
Eventually (future release) it would be good to change wstrings to u16strings because that is ICU native type (I wish it were u32sting, but alas it is not). From what I remember looking at code, the GUI seems to depend upon wstrings being 32-bit (but if that's true, how does it work on Windows).

fbacher commented:
Having some issues with vector<string_view>.
Callers tend to call Split(vector<std::string>, vector<std::string>). 
I have of course defined it as Split(vector<std::string_view>, vector<std::string_view>)
After banging my head a while trying to come up with some elegant Template I went ahead and created another signature
Split(vector<std::string>, vector<std::string), which helped. 
Then one call was: Split(vector<std::string>, {" ", "\t", "\n", "\r"});
Which of course causes the compiler to complain about the ambiguous signatures (since the second arg can either be vector of strings or vector of string_views). 

I can of course fix these, but there ought to be a better way...

garbear commented:
BTW, found your msg hard to read because GB collapses &lt;> (very annoying). You can replace all &lt; with `&lt;` and it'll look normal.

CastagnaIT commented:
where you are using `std::vector<std::string_view>` in the return value of `UnicodeUtils::Split` ?
if so, this is wrong, string_view dont store data in memory it is just a reference to original data, then when you go outside the scope of the method you will have a list of wrong data, because it will point to an invalid memory

if this is not so please explain better by using "code field" button `<>` on text toolbar on example code

fbacher committed:
Change from char to string for string_view

fbacher committed:
Replace now returns bool not int

fbacher committed:
Use string_view where possible

fbacher committed:
Bug in test

Log.FATAL prints critical
Use foldcase option on regex

fbacher committed:
Fix testcase

Use regex fold case option

fbacher committed:
Use string_view, reduce memory footprint

Used stack-based memory instead of malloc based.
Added tests (temporarily) to TestUnicodeUtils for performance and
memory.
Concluded that cputime not significantly impacted by malloc.
Malloc impact significant but not huge.
Saving these changes in a branch, but intend to use simplier, easier to
understand UnicodeString based code.

fbacher commented:
This huge change is understandably difficult to review all at once. I like the suggestion of redoing the changes so they are in smaller chunks and easier to review. 

I have opened another PR to add the ICU library, which this code depends.

I propose that:

- I change one (or several) apis per commit. (Such as toLower)
- I delay moving things from StringUtils to UnicodeUtils 
- At least at first, only commit one thing to this PR at a time

I propose that in the first commit that I change StringUtils.toLower to:

1. use ICU's toLower, where appropriate
2. create StringUtils::FoldCase and switch the appropriate users of toLower over to use that instead.
3. Create Unicode.cpp (like it is in this code drop) to do the real work. It will contain helper functions as well as FoldCase and ToLower.

Please let me know how you think it best to proceed. I have changed the code to use string_view where I can.  I have also made a number of changes to the back-end code (Unicode) to make it easier to review. I also wrote some performance tests to evaluate whether it is worth the extra complexity of using stack-based memory over malloced memory. I have decided that for the most part using the simpler solution is the best way to go. This can be revisited should the situation arise.

fbacher committed:
ICU build

fbacher committed:
Clean up API

Use string_view instead of string
improved comments
Reorg to make clearer and more consistent

fbacher committed:
Merge branch 'master' into UnicodeUtils

fbacher committed:
Update icu build scripts

fbacher committed:
Native build and reduce footprint

Reduced footprint of ICU libs via
options and filters.json to reduce unused icudata

fbacher committed:
Merge branch 'master' into UnicodeUtils

addons/skin.estouchy/language/resource.language.ca_es/strings.po

fbacher committed:
Complete merge

fbacher committed:
Add Performance test
